# Exposing the DNA methylation-responsive compartment of the leukemic genome supports its potential as a novel therapeutic target in T-ALL

Maike Bensberg<sup>1†</sup>, Aida Selimović-Pašić<sup>1†</sup>, Lisa Haglund<sup>1</sup>, Júlia Goldmann<sup>1</sup>, Sandra Hellberg<sup>2†</sup>, Colm E. Nestor<sup>1†</sup>


<sup>1</sup> Crown Princess Victoria Children´s Hospital, Department of Biomedical and Clinical Sciences, Linköping University, Linköping, Sweden.<br />
<sup>2</sup> Division of Inflammation and Infection, Department of Biomedical and Clinical Sciences, Linköping University, Linköping, Sweden.<br />

<sup>†</sup> These authors have contributed equally to this work and share first and last authorship respectively. 

Correspondance to Colm E. Nestor [colm.nestor@liu.se] and Sandra Hellberg [sandra.hellberg@liu.se]

This repository contains the scripts for data analysis and generation of figures presented in [link to paper]

## Data availability

Raw and processed data is available at ArrayExpress: E-MTAB-14375 (total RNA sequencing data), E-MTAB-14202 (high coverage Enzymatic Methyl-sequencing) and E-MTAB-14197 (low-coverage whole methylome sequencing). All other data supporting the findings of this study are available from the corresponding author upon reasonable request.

## Abstract
T-cell acute lymphoblastic leukaemia (T-ALL) exhibits exceptionally high global levels of DNA methylation, with silencing of the DNA demethylating enzyme TET2 implicated in T-ALL’s hypermethylation phenotype. We propose that DNA hypomethylating agents (HMA) could be particularly potent in T-ALL patients with this hypermethylation phenotype. Despite the use of HMAs in treating other haematological malignancies, the role of altered DNA methylation in T-cell cancer biology, including T-ALL, remains unresolved. Here, we used a novel, reversible DNMT1-specific inhibitor and the conventional HMAs, 5-azacytidine and decitabine, to assess the effects of global DNA methylation loss in T-ALL cells and the HMAs’ potential as targeted therapeutic agents in T-ALL. We demonstrate that extensive removal of DNA methylation, even in the absence of DNA damage, results in cell death, and that toxicity is negatively correlated with global levels of DNA methylation. Notably, whereas DNA demethylation caused limited transcriptional changes globally, key tumour suppressor genes, such as TET2 and PAX5, were upregulated in a methylation-dependent manner. Few endogenous retroviruses were reactivated after demethylation, and no profound dysregulation of immune-related genes was observed, challenging the contribution of ‘viral mimicry’ to the observed toxicity. Together, these findings provide fundamental insights into the role of DNA methylation in cancer, clearly demonstrating that the removal of DNA methylation alone is sufficient to induce cell death in cancer cells. Further, the distinct upregulation of silenced tumour suppressor genes has important implications for futur

## :writing_hand: Authors

Sandra Hellberg [sandra.hellberg@liu.se] <br />
Maike Bensberg [maike.bensberg@liu.se] <br />
Linköping University


