# Exposing the DNA methylation-responsive compartment of the leukaemic genome in T-ALL cell lines support its potential as a novel therapeutic target in T-ALL

Maike Bensberg<sup>1†</sup>, Aida Selimović-Pašić<sup>1†</sup>, Lisa Haglund<sup>1</sup>, Júlia Goldmann<sup>1</sup>, Sandra Hellberg<sup>1†</sup>, Colm E. Nestor<sup>1†</sup>


<sup>1</sup> Crown Princess Victoria Children´s Hospital, Department of Biomedical and Clinical Sciences, Linköping University, Linköping, Sweden.<br />

<sup>†</sup> These authors have contributed equally to this work and share first and last authorship respectively. 

Correspondance to Colm E. Nestor [colm.nestor@liu.se] and Sandra Hellberg [sandra.hellberg@liu.se]

This repository contains the scripts for data analysis and generation of figures presented in [link to paper] Accepted for publication in Clinical Epigenetics. 

## Data availability

Raw and processed data is available at ArrayExpress: E-MTAB-14375 (total RNA sequencing data), E-MTAB-14202 (high coverage Enzymatic Methyl-sequencing) and E-MTAB-14197 (low-coverage whole methylome sequencing). All other data supporting the findings of this study are available from the corresponding author upon reasonable request.

## Abstract
T-cell acute lymphoblastic leukaemia (T-ALL) exhibits exceptionally high levels of DNA methylation, with silencing of the DNA demethylating enzyme TET2 implicated in T-ALL’s hypermethylation phenotype. We propose that DNA hypomethylating agents (HMAs) could be particularly potent in T-ALL cells with this phenotype. Here, we used a reversible DNMT1-specific inhibitor and the conventional HMAs, 5-azacytidine and decitabine, to assess the effects of global DNA methylation loss in T-ALL cell lines and the potential of using HMAs as targeted therapeutic agents in T-ALL. We demonstrate that removal of DNA methylation, even in the absence of DNA damage, results in cell death and that toxicity is negatively correlated with methylation levels. Notably, whereas DNA demethylation caused limited transcriptional changes, key tumour suppressor genes, including TET2, were upregulated in a methylation-dependent manner. Few endogenous retroviruses or immune-related genes were reactivated after demethylation, challenging the contribution of ‘viral mimicry’ to HMA toxicity. Together, these findings provide fundamental insights into the role of DNA methylation in T-ALL, demonstrating that the removal of DNA methylation alone is sufficient to (i) induce cell death in T-ALL cell lines and (ii) reactivate silenced tumour suppressor genes. Our findings support the development of therapies targeting the unique methylation phenotype of T-ALL.

## :writing_hand: Authors

Sandra Hellberg [sandra.hellberg@liu.se] <br />
Maike Bensberg [maike.bensberg@liu.se] <br />
Linköping University


