# No evidence for an epigenetic role in the cytotoxicity of traditional DNA demethylating agents in cancer therapy

Maike Bensberg<sup>1†</sup>, Aida Selimović<sup>1†</sup>, Lisa Hedin Haglund<sup>1</sup>, Júlia Goldmann<sup>1</sup>, Sandra Hellberg<sup>2†</sup>, Colm E. Nestor<sup>1†</sup>


<sup>1</sup> Crown Princess Victoria Children´s Hospital, Department of Biomedical and Clinical Sciences, Linköping University, Linköping, Sweden.<br />
<sup>2</sup> Division of Inflammation and Infection, Department of Biomedical and Clinical Sciences, Linköping University, Linköping, Sweden.<br />

<sup>†</sup> These authors have contributed equally to this work and share first and last authorship respectively. 

Correspondance to Colm E. Nestor [colm.nestor@liu.se]

This repository contains the scripts for data analysis and generation of figures presented in [link to paper]

## Data availability

Raw and processed data is available at [add links here]

## Abstract
Irregular profiles of DNA methylation are characteristic for most malignancies and present an accessible therapeutic target. Particularly, the hypomethylating agents 5-azacytidine and 5-aza-2′-deoxycytidine are routinely used for treatment of haematological cancers. These drugs have a known DNA demethylating activity but also cause cytotoxicity through DNA damage. The two differing mechanisms of action confound studies investigating the effect of DNA demethylation in cancer treatment. The novel DNA methyltransferase 1 specific inhibitor GSK-3685032 causes loss of DNA methylation without DNA damage and offers the promising possibility to examine the molecular consequences of global loss of DNA methylation. Here, we use genome-wide low pass methylation sequencing to reveal the demethylating dynamics of all three drugs and show that cytotoxicity of 5-azacytidine and 5-aza-2′-deoxycytidine is primarily due to DNA damage whereas long-term treatment with GSK-3685032 can lower global DNA methylation to less than 30% with limited impacting on cell viability. Combining whole-genome DNA methylation sequencing and total RNA sequencing, we identify DNA methylation-sensitive genes for which expression was directly regulated by DNA methylation. Taken together, our results suggest that HMA induced DNA demethylation in cancer cells is inconsequential to their cytotoxicity, which is mediated entirely through DNA damage.

## Session info

*R version 4.2.0 (2022-04-22)*

| Package       | Version       | Source         |
| ------------- | ------------- |---------------
| `assertthat`      | 0.2.1         | CRAN |
| `BiocParalell`      |  1.29.21  | Bioconductor |
| `corrplot`      | 0.92          | CRAN  |
| `dplyr`         | 1.0.7         | CRAN  |
| `edgeR`     |  3.37.4       | Bioconductor |
| `EnchancedVolcano`     | 1.13.2        | Bioconductor |
| `ggfortify`     | 0.4.14        | CRAN  |
| `ggplot2`       | 3.3.5         | CRAN  |
| `ggpubr`        | 0.4.0         | CRAN  |
| `Glimma`        |  2.5.3      | Bioconductor  |
| `ggrepel`        |  0.9.1      | CRAN |
| `gridExtra`     | 2.3           | CRAN  |
| `hrbrthemes`    | 0.8.0         | CRAN  |
| `limma`    | 3.51.8       | Bioconductor  |
| `MASS`          | 7.3-56        | CRAN  |
| `patchwork`          | 1.1.1        | CRAN  |
| `pheatmap`      | 1.0.12        | CRAN  |
| `ppcor`         | 1.1           | CRAN  |
| `RColorBrewer`         | 1.1-3          | CRAN  |
| `readxl`        | 1.3.1         | CRAN  |
| `statmod`        | 1.4.36        | CRAN  |
| `stringr`       | 1.4.0         | CRAN  |
| `sva`       |   3.15      | Bioconductor |
| `tidyr`       | 1.2.0         | CRAN  |
| `tidyverse`       |    1.3.1      | CRAN  |



## :writing_hand: Authors

Sandra Hellberg [sandra.hellberg@liu.se] <br />
Maike Bensberg [maike.bensberg@liu.se] <br />
Linköping University


