# Exposing the DNA methylation-responsive compartment of the leukemic transcriptome questions its relevance as a therapeutic target

Maike Bensberg<sup>1†</sup>, Aida Selimović-Pašić<sup>1†</sup>, Lisa Haglund<sup>1</sup>, Júlia Goldmann<sup>1</sup>, Sandra Hellberg<sup>2†</sup>, Colm E. Nestor<sup>1†</sup>


<sup>1</sup> Crown Princess Victoria Children´s Hospital, Department of Biomedical and Clinical Sciences, Linköping University, Linköping, Sweden.<br />
<sup>2</sup> Division of Inflammation and Infection, Department of Biomedical and Clinical Sciences, Linköping University, Linköping, Sweden.<br />

<sup>†</sup> These authors have contributed equally to this work and share first and last authorship respectively. 

Correspondance to Colm E. Nestor [colm.nestor@liu.se] and Sandra Hellberg [sandra.hellberg@liu.se]

This repository contains the scripts for data analysis and generation of figures presented in [link to paper]

## Data availability

Raw and processed data is available at ArrayExpress: E-MTAB-14375 (total RNA sequencing data), E-MTAB-14202 (high coverage Enzymatic Methyl-sequencing) and E-MTAB-14197 (low-coverage whole methylome sequencing). All other data supporting the findings of this study are available from the corresponding author upon reasonable request.

## Abstract
**Background**: Mutations in proteins that regulate DNA methylation are implicated in the development and progression of several cancers, and aberrant DNA methylation is a universal feature of neoplasia. Despite the current use of DNA hypomethylating agents (HMAs) in treating haematological malignancies, the role of altered DNA methylation patterns in cancer biology remains largely unresolved. 

**Methods**: In this study, we used a novel, reversible DNMT1-specific inhibitor as well as conventional HMAs to assess the molecular effects of global DNA methylation loss in leukemic cells, integrating toxicity studies, whole genome DNA methylation and total RNA-sequencing.

**Results**: We demonstrate that removal of DNA methylation has minimal impact on cell viability and does not generally upregulate transcription. Furthermore, few endogenous retroviruses become expressed following demethylation, challenging the concept of ‘viral mimicry’ as a source of toxicity. While demethylation led to the reactivation of a handful of tumour suppressor genes including TET2, it resulted in the expression of twice as many oncogenes clearly implicated in leukemic progression, such as TAL1 and TLX1. Indeed, this study revealed no evidence of an epigenetic basis for the toxicity associated with FDA-approved HMAs. 

**Conclusions**: Together, these findings offer a nuanced understanding of the interplay between DNA methylation with transcription and cytotoxicity in cancer and have important implications for future therapeutic strategies targeting aberrant DNA methylation.

## :writing_hand: Authors

Sandra Hellberg [sandra.hellberg@liu.se] <br />
Maike Bensberg [maike.bensberg@liu.se] <br />
Linköping University


