### Pre-processing RNA-seq analysis of count data generated by TElocal

# Title: TElocal_RNAseq_SUPT1_TE

# Author: Sandra Hellberg
# Email: sandra.hellberg@liu.se

# Date: 2024-01-24
# Last modified: 2024-08-06

# Description: Total RNA-seq data has been pre-processed including quality control,
# trimming, alignment (STAR), quantification with TElocal. The script contains
# the downstream analysis based on the RNAseq123. 

# Comment: https://www.bioconductor.org/packages/devel/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html#reading-in-count-data 
# Comment: https://github.com/mhammell-laboratory/TEtranscripts 
# Comment: https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf

# Samples: 4 samples Loucy, 4 samples SUP-T1 (the two cell lines are processed separately)

rm(list=ls()) # remove all entries in the global environment 
set.seed(301) # seed for reproducibility

#-------------------------------------------------------------------------------

            ### 0. Set directory structure and load packages 

#-------------------------------------------------------------------------------
setwd("working directory")
main_dir <- "input directory"

lib_path <- .libPaths("C:/Rpackages/") # specifing the package library
#update.packages(lib.loc = lib_path, ask = FALSE, checkBuilt = TRUE, dependencies = TRUE)
#BiocManager::install(lib.loc = lib_path, update = TRUE, ask = FALSE)

pack_R <- c("limma","edgeR", "dplyr", "tidyverse", "stats")

for (i in 1:length(pack_R)) {
  library(pack_R[i], character.only = TRUE)
}

#-------------------------------------------------------------------------------

                    ### 1. Read in count data-----

#-------------------------------------------------------------------------------
input <- list.files(pattern=".cntTable") # list all files, should be 8 in total
stopifnot(length(input)==8)

input_names <- input %>% # create a vector with the file names to be used in the for loop
  str_remove("TElocal_240704_") %>%
  str_remove("_Aligned.cntTable") 

for(i in 1:length(input)) {
  x <- read.delim(paste0(main_dir, input[i]))
  colnames(x) <- c("geneid", input_names[i])
  assign(input_names[i], x)
  rm(x)
}

# check if the rows are in the same order in all data frames
files <- list(i2_L_D10_3d_totalRNA_S2, i3_L_G300_3d_totalRNA_concat_S1_2,
              i5_L_G300_7d_totalRNA_S4, i7_S_D10_3d_totalRNA_S5, i8_S_G300_3d_totalRNA_S6, i9_S_control_3d_totalRNA_S7,i10_S_G300_7d_totalRNA_S8)

for(i in 1:length(files)) {
  if(all.equal(rownames(i4_L_control_3d_totalRNA_S3), rownames(files[[i]]))) {
    print("The order of rows is the same")
  } else {
    print("The order of rows is different")
  }
}

#-------------------------------------------------------------------------------

                      ### 2. Format count data-----

#-------------------------------------------------------------------------------
# SUP-T1 (i9, i7, i8, i10)
SUPT1_list <- list(i9_S_control_3d_totalRNA_S7, i7_S_D10_3d_totalRNA_S5, i8_S_G300_3d_totalRNA_S6, i10_S_G300_7d_totalRNA_S8)
SUPT1_counts <- data.frame(SUPT1_list %>% reduce(full_join, by='geneid')) # merge to one count table

# Read in ribosomal RNA annotation and remove from count data before downstream analysis
rRNA <- read.delim("rRNA_annotation.txt")
rRNA <- rRNA[,2]
SUPT1_counts <- SUPT1_counts[!SUPT1_counts$geneid %in% rRNA,] # 4741209-4741170 = 39 rRNA removed from the data

raw_counts <- SUPT1_counts 
rownames(raw_counts) <- raw_counts$geneid
raw_counts$geneid <- NULL
raw_counts <- as.matrix(raw_counts)

rownames_genes <- rownames(raw_counts) # all genes/TEs present in the count data
stopifnot(nrow(raw_counts)==length(rownames_genes))

# Select TEs to include (ERV=LTR, LINE = L1, L2, SINE = Alu, MIR)
rownames_TE <- grep(":", rownames(raw_counts), value = TRUE, fixed = TRUE) # All TEs have : in their names and can be used to select all TEs

# ERVs
rownames_LTR <- grep("LTR", rownames_TE, value = TRUE)
rownames_ERV <- grep("ERV", rownames_TE, value = TRUE)
rownames_ERV <- unique(c(rownames_LTR, rownames_ERV)) # remove duplicated entries if any

# LINEs
rownames_LINE <- grep("LINE", rownames_TE, value = TRUE)
rownames_L1 <- grep("L1", rownames_LINE, value = TRUE)
rownames_L2 <- grep("L2", rownames_LINE, value = TRUE)
rownames_LINE <- c(rownames_L1, rownames_L2)

# SINEs
rownames_SINE <- grep("SINE", rownames_TE, value = TRUE)
rownames_Alu <- grep("Alu", rownames_SINE, value = TRUE)
rownames_MIR <- grep("MIR", rownames_SINE, value = TRUE)
rownames_SINE <- c(rownames_Alu, rownames_MIR)

rownames_ERV_LINE_SINE <- c(rownames_ERV, rownames_LINE, rownames_SINE)
sum((grepl(":", rownames_ERV_LINE_SINE))) # ensure that only rows with : are included 

# subset the full count matrix to only contain the selected TEs
input_data_TE <- raw_counts[rownames(raw_counts) %in% rownames_ERV_LINE_SINE, , drop = FALSE] # ERV, LINE & SINE
stopifnot(nrow(input_data_TE) == length(rownames_ERV_LINE_SINE))

input_data <- input_data_TE # TE raw count matrix

#-------------------------------------------------------------------------------

                ### 3. Create and format DGElist------

#---------------------------------------------------------------------------------
# create a DGE list
counts <- DGEList(input_data) # create a DGE list
class(counts)
dim(counts)

# Organizing sample information
colnames(counts)
group <- as.factor(c("control_3", "D10_3", "GSK300_3", "GSK300_7"))
counts$samples$group <- group # add group to DGElist

# Removing genes that are lowly expressed
table(rowSums(counts$counts==0)==4) # check the number of rows where the count is 0 for all samples

dim(counts) # check before removing lowly expressed genes
keep.exprs <- filterByExpr(counts, group=group)
counts <- counts[keep.exprs,, keep.lib.sizes=FALSE]
dim(counts) # check after removing lowly expressed genes

counts

write.table(counts$counts, paste0(main_dir, "SUPT1_RNAseq_rawcounts_TE_240806.txt"), sep = "\t", row.names = T)

# CPM and LCPM
cpm <- cpm(counts)
write.table(cpm, paste0(main_dir, "SUPT1_RNAseq_cpm_TE_240806.txt"), sep = "\t", row.names = T)

lcpm <- cpm(counts, log=TRUE)
write.table(lcpm, paste0(main_dir, "SUPT1_RNAseq_lcpm_TE_240806.txt"), sep = "\t", row.names = T)


#-------------------------------------------------------------------------------

                  ### 4. Differential analysis-----

#-------------------------------------------------------------------------------
# based on the edgeR manual 2.12: What to do if you do not have replicates
# recommended dispersion parameter for controlled human data is 0.4. 

counts$samples$group # check the order of the samples because it was exactTest will use for the comparision

bcv = 0.4 

# DEC versus control
DEC_Ctrl <- exactTest(counts, pair=c(1, 2), dispersion = bcv^2) # 1 = control_3 , 2 = D10_3
DEC_Ctrl_test <- DEC_Ctrl$table
hist(DEC_Ctrl_test$PValue)
DEC_Ctrl_test$padjust <- p.adjust(DEC_Ctrl_test$PValue, method = "BH") # add adjusted pvalues
nrow(DEC_Ctrl_test[DEC_Ctrl_test$padjust < 0.05 & DEC_Ctrl_test$logFC > 1, ]) # number of up-regulated genes

write.table(DEC_Ctrl_test, paste0(main_dir, "SUPT1_DECvsCtrl_RNAseq_exactTest_TE_bcv0.4_240806.txt"), sep = "\t", row.names = T)


# GSK_3 versus control
GSK_3_Ctrl <- exactTest(counts, pair=c(1, 3), dispersion = bcv^2) # 1 = control_3, 3 = GSK300_3
GSK_3_Ctrl_test <- GSK_3_Ctrl$table
hist(GSK_3_Ctrl_test$PValue)
GSK_3_Ctrl_test$padjust <- p.adjust(GSK_3_Ctrl_test$PValue, method = "BH") # add adjusted pvalues
nrow(GSK_3_Ctrl_test[GSK_3_Ctrl_test$padjust < 0.05 & GSK_3_Ctrl_test$logFC > 1 ,]) # number of up-regulated TEs

write.table(GSK_3_Ctrl_test, paste0(main_dir, "SUPT1_GSK3vsCtrl_RNAseq_exactTest_TE_bcv0.4_240806.txt"), sep = "\t", row.names = T)

# GSK_7 versus control
GSK_7_Ctrl <- exactTest(counts, pair=c(1, 4), dispersion = bcv^2) # 1 = control_3, 4 = GSK300_7
GSK_7_Ctrl_test <- GSK_7_Ctrl$table
hist(GSK_7_Ctrl_test$PValue)
GSK_7_Ctrl_test$padjust <- p.adjust(GSK_7_Ctrl_test$PValue, method = "BH") # add adjusted pvalues
nrow(GSK_7_Ctrl_test[GSK_7_Ctrl_test$padjust < 0.05 & GSK_7_Ctrl_test$logFC > 1, ]) # number of up-regulated TEs

write.table(GSK_7_Ctrl_test, paste0(main_dir, "SUPT1_GSK7vsCtrl_RNAseq_exactTest_TE_bcv0.4_240806.txt"), sep = "\t", row.names = T)


## SCRIPT STOPS HERE

